File: ratemyprofessors/src/main/java/com/tevinjeffrey/rmp/scraper/RMPScraper.java
package com.tevinjeffrey.rmp.scraper;

import com.tevinjeffrey.rmp.common.Parameter;
import com.tevinjeffrey.rmp.common.Professor;
import com.tevinjeffrey.rmp.scraper.search.Decider;
import com.tevinjeffrey.rmp.scraper.search.RatingParser;
import com.tevinjeffrey.rmp.scraper.search.SearchParser;

import java.io.IOException;

import javax.inject.Inject;

import io.reactivex.Observable;
import io.reactivex.ObservableOnSubscribe;
import io.reactivex.schedulers.Schedulers;
import okhttp3.Call;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;

import static com.tevinjeffrey.rmp.common.RMP.RMP_BASE_URL;

public class RMPScraper {

  final static String TAG = RMPScraper.class.getSimpleName();

  OkHttpClient client;

  @Inject
  public RMPScraper(OkHttpClient client) {
    this.client = client;
  }

  public String createUrlFromParams(Parameter params) {
    return RMP_BASE_URL + "/search.jsp?stateselect=nj&query=" + params.lastName +
        (params.university.toLowerCase().contains("rutgers") ? "+rutgers" : "");
  }

  public Observable<Professor> findBestProfessor(final Parameter params) {
    Observable<? extends Professor> firstChoice = getProfessors(params);
    params.university = "";
    Observable<? extends Professor> secondChoice = getProfessors(params);

    return Observable.concat(firstChoice, secondChoice)
        .firstElement().toObservable();
  }

  public Observable<? extends Professor> getProfessors(final Parameter params) {
    return performSearch(createUrlFromParams(params))
        .flatMap(this::getProfessors)
        .filter(professor -> professor != null)
        .toList()
        .toObservable()
        .flatMap(professors -> Observable.fromIterable(Decider.determineProfessor(
            professors,
            params
        )));
  }

  public Observable<String> makeRequest(String url, String tag) {
    return makeGetCall(new Request.Builder()
        .tag(tag)
        .url(url)
        .build())
        .flatMap(this::executeGetCall)
        .flatMap(this::mapResponseToString);
  }

  public Observable<String> performSearch(final String url) {
    return makeRequest(url, TAG)
        .flatMap(initalSearch -> {
          final int numberOfProfessors = SearchParser.getNumberOfProfessors(initalSearch);
          if (numberOfProfessors <= 20) {
            return Observable.just(initalSearch);
          } else if (numberOfProfessors > 20) {
            return Observable.create((ObservableOnSubscribe<String>) e -> {
              if (!e.isDisposed()) {
                for (int i = 0; i <= numberOfProfessors; i += 20) {
                  e.onNext(url + "&offset=" + i);
                }
                e.onComplete();
              }
            }).flatMap(s -> makeRequest(url, TAG));
          } else {
            return Observable.empty();
          }
        });
  }

  private Observable<Response> executeGetCall(final Call call) {
    return Observable.create(e -> {
      if (e != null && !e.isDisposed()) {
        try {
          e.onNext(call.execute());
          e.onComplete();
        } catch (IOException thowable) {
          e.onError(thowable);
        }
      }
    });
  }

  private Observable<ScrapeProfessor> getProfessors(String response) {
    return Observable.fromIterable(SearchParser.getSearchResults(response))
        .filter(listing -> !listing.getUrl().contains("Add"))
        .flatMap(listing -> makeRequest(RMP_BASE_URL + listing.getUrl(), TAG)
            .flatMap(s -> Observable.just(RatingParser.findProfessor(listing, s)))
            .subscribeOn(Schedulers.io()));
  }

  private Observable<Call> makeGetCall(final Request request) {
    return Observable.create(e -> {
      if (e != null && !e.isDisposed()) {
        e.onNext(client.newCall(request));
        e.onComplete();
      }
    });
  }

  private Observable<String> mapResponseToString(final Response response) {
    System.out.println(response.request().toString());
    return Observable.create(e -> {
      if (e != null && !e.isDisposed()) {
        try {
          e.onNext(response.body().string());
          e.onComplete();
        } catch (IOException throwable) {
          e.onError(throwable);
        }
      }
    });
  }
}

File: ratemyprofessors/src/main/java/com/tevinjeffrey/rmp/scraper/RMPScraper.kt
package com.tevinjeffrey.rmp.scraper

import com.tevinjeffrey.rmp.common.Parameter
import com.tevinjeffrey.rmp.common.Professor
import com.tevinjeffrey.rmp.common.RMP.Companion.RMP_BASE_URL
import com.tevinjeffrey.rmp.scraper.search.Decider
import com.tevinjeffrey.rmp.scraper.search.RatingParser
import com.tevinjeffrey.rmp.scraper.search.SearchParser
import io.reactivex.Observable
import io.reactivex.schedulers.Schedulers
import okhttp3.Call
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import java.io.IOException
import javax.inject.Inject

class RMPScraper @Inject
constructor(internal var client: OkHttpClient) {

  private fun createUrlFromParams(params: Parameter): String {
    return RMP_BASE_URL + "/search.jsp?stateselect=nj&query=" + params.lastName +
        if (params.university!!.toLowerCase().contains("rutgers")) "+rutgers" else ""
  }

  fun findBestProfessor(params: Parameter): Observable<Professor> {
    val firstChoice = getProfessors(params)
    params.university = ""
    val secondChoice = getProfessors(params)

    return Observable.concat(firstChoice, secondChoice)
        .firstElement().toObservable()
  }

  fun getProfessors(params: Parameter): Observable<out Professor> {
    return performSearch(createUrlFromParams(params))
        .flatMap({
          this.getProfessors(it)
        })
        .toList()
        .toObservable()
        .flatMap { professors ->
          Observable.fromIterable(Decider.determineProfessor(
              professors,
              params
          ))
        }
  }

  private fun makeRequest(url: String, tag: String): Observable<String> {
    return makeGetCall(Request.Builder()
        .tag(tag)
        .url(url)
        .build())
        .flatMap({ this.executeGetCall(it) })
        .flatMap({ this.mapResponseToString(it) })
  }

  private fun performSearch(url: String): Observable<String> {
    return makeRequest(url, TAG)
        .flatMap { initalSearch ->
          val numberOfProfessors = SearchParser.getNumberOfProfessors(initalSearch)
          when {
            numberOfProfessors <= 20 -> {
              Observable.just(initalSearch)
            }
            numberOfProfessors > 20 -> {
              Observable.create<String>({ e ->
                if (!e.isDisposed) {
                  var i = 0
                  while (i <= numberOfProfessors) {
                    e.onNext(url + "&offset=" + i)
                    i += 20
                  }
                  e.onComplete()
                }
              })
                  .flatMap { s -> makeRequest(s, s) }
            }
            else -> Observable.empty()
          }
        }
  }

  private fun executeGetCall(call: Call): Observable<Response> {
    return Observable.create { e ->
      if (!e.isDisposed) {
        try {
          e.onNext(call.execute())
          e.onComplete()
        } catch (thowable: IOException) {
          e.onError(thowable)
        }
      }
    }
  }

  private fun getProfessors(response: String): Observable<Professor> {
    return Observable.fromIterable(SearchParser.getSearchResults(response))
        .filter { listing -> !listing.url.contains("Add") }
        .flatMap { listing ->
          makeRequest(RMP_BASE_URL + listing.url, TAG)
              .flatMap { s -> Observable.just(RatingParser.findProfessor(listing, s)) }
              .subscribeOn(Schedulers.io())
        }
  }

  private fun makeGetCall(request: Request): Observable<Call> {
    return Observable.create { e ->
      if (!e.isDisposed) {
        e.onNext(client.newCall(request))
        e.onComplete()
      }
    }
  }

  private fun mapResponseToString(response: Response): Observable<String> {
    println(response.request().toString())
    return Observable.create { e ->
      if (!e.isDisposed) {
        try {
          e.onNext(response.body()!!.string())
          e.onComplete()
        } catch (throwable: IOException) {
          e.onError(throwable)
        }
      }
    }
  }

  companion object {
    internal val TAG = RMPScraper::class.java.simpleName
  }
}
